### PR TITLE
Complete Network Functionality

### PR DESCRIPTION
**Pull request by claywatkins**
This PR contains complete network functionality with our web based back end. 

This includes: 
- Getting city details
- Saving cities to favorites
- Retrieving all favorited cities
- Retrieving city data without a name
- Deletion of cities from web based back end

Also included is all the relevant updates to the UI to display all returned/changed data. Please keep in mind that all UI will need to be updated as it's current state is reflected with only placeholders.

### CODE REVIEW

File: labs-ios-starter/Model Controller/LocationController.swift [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                let returnedLocation = try JSONDecoder().decode(Location.self, from: data)
+                completion(.success(returnedLocation))
+            } catch {
+                print("Error decoding JSON: \(error)")
+                completion(.failure(.badDecode))
+            }
+        }
+        task.resume()
+    }
+
```
Comment (iOSjess) [inline labs-ios-starter/Model Controller/LocationController.swift:156]: Did you figure out what we need this for?

File: labs-ios-starter/Model Controller/ProfileController.swift [hunk 2]
```diff
@@ -52,7 +54,8 @@ class ProfileController {
         var request = URLRequest(url: requestURL)
         
         request.addValue("Bearer \(oktaCredentials.idToken)", forHTTPHeaderField: "Authorization")
-        
+        self.bearer = oktaCredentials.idToken
+        print(bearer)
```
Comment (iOSjess) [inline labs-ios-starter/Model Controller/ProfileController.swift]: We might want to remove this print statement.

File: labs-ios-starter/Model/Location.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    var state: String?
+    
+    enum CodingKeys: String, CodingKey{
+        case id = "id_num"
+        case crimeRate = "crime_rate"
+        case rentalRate = "rental_rate"
+        case walkScore = "walk_score"
+        case city, state, population
+    }
+}
```
Comment (iOSjess) [inline labs-ios-starter/Model/Location.swift:27]: This is good. Clean and clear.

File: labs-ios-starter/Storyboards/Map/SelectedLocationDetailViewController.swift [hunk 4]
```diff
@@ -9,22 +9,55 @@
...
+    @IBAction func saveToFavoritesTapped(_ sender: UIButton) {
+        guard let name = locationName else { return }
+        guard let id = locationID else { return }
+        let savedcity = SavedLocation(name: name, cityID: id)
+        locationController.saveCityAsFavorite(location: savedcity) { (_) in
+            DispatchQueue.main.async {
+                let ac = UIAlertController(title: "Location Saved", message: "\(name) has been added to favorites.", preferredStyle: .alert)
+                ac.addAction(UIAlertAction(title: "Okay", style: .default, handler: nil))
+                self.present(ac, animated: true)
+            }
```
Comment (iOSjess) [inline labs-ios-starter/Storyboards/Map/SelectedLocationDetailViewController.swift:59]: Good. I love these alerts. They play an important role in the user experience.

**Review by iOSjess — APPROVED** [review]
Looks like you got a lot done. It's nice to know we can use the data now.

File: labs-ios-starter/Model Controller/ProfileController.swift (hunk 2 above)
Comment (claywatkins) [reply labs-ios-starter/Model Controller/ProfileController.swift]: Print statement has now been removed!

**Pull request merged**